### ISSUE TITLE
[LFRic] [PSyAD] Missing halo exchange and immediate cleaning of halos when using setval_random with redundant calculations

### ISSUE THREAD

**Issue by DrTVockerodtMO** [stfc/PSyclone#2805]
We are testing some LFRic branches that run the adjoint tests in parallel. We find that when using `setval_random` there is a missing halo exchange in the PSy layer code after the field values have been randomised. This causes some failures in adjoint tests that would otherwise pass if the halo is properly initialised.

With redundant calculations, the halo exchange would be missed because after the field is randomised, the halo is set to dirty but then immediately set to clean. This isn't present when using `setval_random` normally. However, there is also no halo exchange after `setval_random` is used before it enters the forward kernels, which when run in parallel causes floating invalid errors.

Ultimately, we need to perform a halo exchange between the use of `setval_random` and the forward kernel in order to get the tests to work properly, whilst still randomising the annexed dofs.

As an example, looking at `atlt_vorticity_advection_alg_mod` we have the following invoke:

```
call invoke( setval_random(r_u), setval_x(r_u_input, r_u), &
             setval_random(wind), setval_x(wind_input, wind), &
             setval_random(vorticity), setval_x(vorticity_input, vorticity), &
             setval_random(ls_wind), setval_random(ls_vorticity), &
             tl_vorticity_advection_kernel_type(r_u, wind, vorticity, ls_wind, ls_vorticity, chi, panel_id, qr_xyoz), &
             x_innerproduct_x(r_u_inner_prod, r_u), &
             x_innerproduct_x(wind_inner_prod, wind), &
             x_innerproduct_x(vorticity_inner_prod, vorticity) )
```

which when compiled using LFRic parallel optimisations and the options described above yields

```
      !
      ! Set-up all of the loop bounds
      !
      loop0_start = 1
      loop0_stop = r_u_proxy%vspace%get_last_dof_halo(1)
      loop1_start = 1
      loop1_stop = r_u_input_proxy%vspace%get_last_dof_annexed()
      loop2_start = 1
      loop2_stop = wind_proxy%vspace%get_last_dof_halo(1)
      loop3_start = 1
      loop3_stop = wind_input_proxy%vspace%get_last_dof_annexed()
      loop4_start = 1
      loop4_stop = vorticity_proxy%vspace%get_last_dof_halo(1)
      loop5_start = 1
      loop5_stop = vorticity_input_proxy%vspace%get_last_dof_annexed()
      loop6_start = 1
      loop6_stop = ls_wind_proxy%vspace%get_last_dof_halo(1)
      loop7_start = 1
      loop7_stop = ls_vorticity_proxy%vspace%get_last_dof_halo(1)
      loop8_start = 1
      loop8_stop = ncolour
      loop9_start = 1
      loop10_start = 1
      loop10_stop = r_u_proxy%vspace%get_last_dof_owned()
      loop11_start = 1
      loop11_stop = wind_proxy%vspace%get_last_dof_owned()
      loop12_start = 1
      loop12_stop = vorticity_proxy%vspace%get_last_dof_owned()
      !
      ! Call kernels and communication routines
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop0_start,loop0_stop
        CALL RANDOM_NUMBER(r_u_data(df))
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL r_u_proxy%set_dirty()
      CALL r_u_proxy%set_clean(1)
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop1_start,loop1_stop
        r_u_input_data(df) = r_u_data(df)
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL r_u_input_proxy%set_dirty()
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop2_start,loop2_stop
        CALL RANDOM_NUMBER(wind_data(df))
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL wind_proxy%set_dirty()
      CALL wind_proxy%set_clean(1)
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop3_start,loop3_stop
        wind_input_data(df) = wind_data(df)
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL wind_input_proxy%set_dirty()
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop4_start,loop4_stop
        CALL RANDOM_NUMBER(vorticity_data(df))
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL vorticity_proxy%set_dirty()
      CALL vorticity_proxy%set_clean(1)
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop5_start,loop5_stop
        vorticity_input_data(df) = vorticity_data(df)
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL vorticity_input_proxy%set_dirty()
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop6_start,loop6_stop
        CALL RANDOM_NUMBER(ls_wind_data(df))
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL ls_wind_proxy%set_dirty()
      CALL ls_wind_proxy%set_clean(1)
      !
      ! End of set dirty/clean section for above loop(s)
      !
      !$omp parallel default(shared), private(df)
      !$omp do schedule(static)
      DO df=loop7_start,loop7_stop
        CALL RANDOM_NUMBER(ls_vorticity_data(df))
      END DO
      !$omp end do
      !$omp end parallel
      !
      ! Set halos dirty/clean for fields modified in the above loop(s)
      !
      CALL ls_vorticity_proxy%set_dirty()
      CALL ls_vorticity_proxy%set_clean(1)
      !
      ! End of set dirty/clean section for above loop(s)
      !
      IF (chi_proxy(1)%is_dirty(depth=1)) THEN
        CALL chi_proxy(1)%halo_exchange(depth=1)
      END IF
      !
      IF (chi_proxy(2)%is_dirty(depth=1)) THEN
        CALL chi_proxy(2)%halo_exchange(depth=1)
      END IF
      !
      IF (chi_proxy(3)%is_dirty(depth=1)) THEN
        CALL chi_proxy(3)%halo_exchange(depth=1)
      END IF
      !
      IF (panel_id_proxy%is_dirty(depth=1)) THEN
        CALL panel_id_proxy%halo_exchange(depth=1)
      END IF
      !
      DO colour=loop8_start,loop8_stop
        !$omp parallel default(shared), private(cell)
        !$omp do schedule(static)
        DO cell=loop9_start,last_halo_cell_all_colours(colour,1)
          !
          CALL tl_vorticity_advection_code(nlayers, r_u_data, wind_data, vorticity_data, ls_wind_data, ls_vorticity_data, &
&chi_1_data, chi_2_data, chi_3_data, panel_id_data, ndf_w2, undf_w2, map_w2(:,cmap(colour,cell)), basis_w2_qr_xyoz, ndf_w1, &
&undf_w1, map_w1(:,cmap(colour,cell)), basis_w1_qr_xyoz, ndf_aspc9_chi, undf_aspc9_chi, map_aspc9_chi(:,cmap(colour,cell)), &
&basis_aspc9_chi_qr_xyoz, diff_basis_aspc9_chi_qr_xyoz, ndf_adspc3_panel_id, undf_adspc3_panel_id, &
&map_adspc3_panel_id(:,cmap(colour,cell)), np_xy_qr_xyoz, np_z_qr_xyoz, weights_xy_qr_xyoz, weights_z_qr_xyoz)
        END DO
        !$omp end do
        !$omp end parallel
      END DO
      !
      ! Set halos dirty/clean for fields modified in the above loop
      !
      CALL r_u_proxy%set_dirty()
      !
      ... ...
```

where I have omitted the inner product calculations for brevity. Running this as-is causes a floating invalid error in the forward code.

**Comment by arporter** [comment]
`setval_random` by itself should not trigger a halo exchange, this should be performed before whatever *reads* the field that has been written to by that kernel. Is the metadata for the subsequent kernel correct? (Performing redundant computation in `setval_random` might then remove this subsequent halo exchange BTW.)

**Comment by DrTVockerodtMO** [comment]
> `setval_random` by itself should not trigger a halo exchange, this should be performed before whatever _reads_ the field that has been written to by that kernel. Is the metadata for the subsequent kernel correct? (Performing redundant computation in `setval_random` might then remove this subsequent halo exchange BTW.)

I believe the metadata is fine because the next part of the invokes are the forward kernels. I just had a look at some PSy layer code and remembered the actual problem so I will edit the description above.

**Comment by arporter** [comment]
There must be something special going on here as, from a PSyclone point of view, there's nothing special about `setval_random` c.f. the other `setval_*` builtins. They've been in use for a long time with no problems reported about missing halo exchanges. However, you mention 'annexed dofs' and that might be relevant. If you are running PSyclone with the 'annexed dofs' option set to True (https://psyclone.readthedocs.io/en/latest/dynamo0p3.html#annexed-dofs) then all builtins will (or should) compute out to the L1 halo to ensure that annexed dofs are clean. Does that shed any light on the situation?

**Comment by arporter** [comment]
Is this a new issue or was it present in the 2.5.0 release do you know?

**Comment by DrTVockerodtMO** [comment]
> Is this a new issue or was it present in the 2.5.0 release do you know? (I could have broken something in the recent work I did on the halo-exchange logic.)

I am not entirely certain I'm afraid, sorry!

**Comment by arporter** [comment]
YGM on Teams.

**Comment by TeranIvy** [comment]
Hi all! @DrTVockerodtMO, I am not sure what version of PSyclone you are using. Is it test environment or 2.5.0? I modified test environment for Josh's testing of operator changes. If you are using test environment, what is the LFRic branch you are using? In any case, I will reinstall PSyclone test environment from master and I would suggest double-checking that everything works.

**Comment by TeranIvy** [comment]
@arporter, as far as I can see the current LFRic Core and Apps are fine with 2.5.0.

**Comment by DrTVockerodtMO** [comment]
> Hi all! @DrTVockerodtMO, I am not sure what version of PSyclone you are using. Is it test environment or 2.5.0? I modified test environment for Josh's testing of operator changes. If you are using test environment, what is the LFRic branch you are using? In any case, I will reinstall PSyclone test environment from master and I would suggest double-checking that everything works.

This is not the test environment no.

**Comment by arporter** [comment]
The metadata for matrix-vector is:
```
  type, public, extends(kernel_type) :: matrix_vector_kernel_type
    private
    type(arg_type) :: meta_args(3) = (/                                    &
         arg_type(GH_FIELD,    GH_REAL, GH_INC,  ANY_SPACE_1),             &
         arg_type(GH_FIELD,    GH_REAL, GH_READ, ANY_SPACE_2),             &
         arg_type(GH_OPERATOR, GH_REAL, GH_READ, ANY_SPACE_1, ANY_SPACE_2) &
         /)
    integer :: operates_on = CELL_COLUMN
  end type
```
so it *does* perform GH_INC on a field on a continuous (or potentially continuous) function space. This will require annexed dofs to be clean on entry. This would normally trigger a halo exchange but, when using `COMPUTE_ANNEXED_DOFS` (s the MO do by default), the annexed dofs are always assumed to be clean and no halo exchange is added. The annexed dofs are made clean by ensuring that the `setval_*` kernels always (redundantly) update them. However, because matrix-vector operates on cells, it will still be accessing 'halo' dofs, even if the results are never used. Therefore, if those 'halo' dofs contain invalid data, we will get f.p. exceptions. The recommended way to ensure this doesn't happen is to always perform redundant computation to the L1 halo for *all* `setval_` kernels. Once we see the generated PSy-layer, we'll be able to see definitively whether this is being done.

**Comment by arporter** [comment]
Which version of PSyclone are you using @DrTVockerodtMO? I would have expected to see comments in the PSy-layer identifying each of the builtins? The problem seems to be that some of the builtins are only doing redundant updates for the annexed dofs rather than the whole L1 halo. This then means the GH_INC access in the kernel is accessing uninitialised data.

**Comment by DrTVockerodtMO** [comment]
> Which version of PSyclone are you using @DrTVockerodtMO? I would have expected to see comments in the PSy-layer identifying each of the builtins? The problem seems to be that some of the builtins are only doing redundant updates for the annexed dofs rather than the whole L1 halo. This then means the GH_INC access in the kernel is accessing uninitialised data.

I am using PSyclone 2.5.0.

**Comment by mo-joshuacolclough** [comment]
> The recommended way to ensure this doesn't happen is to always perform redundant computation to the L1 halo for all setval_ kernels

It would seem to me that `setval_random` is a special case here - each MPI PE is not guaranteed to generate values which match the "owned" dofs on another partition, due to the order in which the random numbers are generated (this will naturally change as the partitioning & indexing changes). Perhaps there is a clever way to account for this somehow? But if not then the only way these match would be with a halo exchange, otherwise the halo will contain "dirty" halo information.

**Comment by mo-joshuacolclough** [comment]
With the (non-stencil) adjoint tests, if the halo contains "garbage" data then it doesn't matter so much, so long as the test runs and passes the adjoint on the owned dofs. It's worth bearing in mind that in reality when the adjoint model is run, the halos should already be populated as part of the application run (a read to/from file/JEDI will have happened by the time the routines are called). If this was not the case we would be seeing issues in the linear model work.

A potential fix it seems to me would be to edit the PSyAD test harness generation to ensure the halo values are at least assigned to something. Be that with a `halo_exchange` or through using `setval_c` to set the redundant halo points to an arbitrary value.

**Comment by arporter** [comment]
> 
> It would seem to me that `setval_random` is a special case here - each MPI PE is not guaranteed to generate values which match the "owned" dofs on another partition, due to the order in which the random numbers are generated (this will naturally change as the partitioning & indexing changes). Perhaps there is a clever way to account for this somehow? But if not then the only way these match would be with a halo exchange, otherwise the halo will contain "dirty" halo information. Sorry that this doesn't help address the main issue here much...

That's a very good point which I hadn't considered: essentially it's not possible to do redundant 'computation' when creating pseudo random numbers (well, I guess it is possible but it would be quite tricky). Have you tried turning off the 'annexed dofs' configuration option?

**Comment by mo-joshuacolclough** [comment]
Thanks @arporter I will attempt it and update with what happens.

**Comment by mo-joshuacolclough** [comment]
Interestingly it gets further (two adjoint tests complete), but it fails on the third test with a `floating invalid` error once again, this time in the adjoint side of the test.

**Comment by arporter** [comment]
Hmm, is the metadata on the adjoint kernel correct?

**Comment by DrTVockerodtMO** [comment]
I had a discussion with @arporter and it looks like setting `COMPUTE_ANNEXED_DOFS` to false in the `psyclone.cfg` fixes this. I will do some more testing to see the differences with it on and off.

I have a branch in LFRic apps open which plumbs in a custom `psyclone.cfg` specifically for the `adjoint_tests` application since this is the only thing currently that uses `setval_random`. The `adjoint` itself by default will use the global config so that the other builtins are not slowed down by unnecessary halo exchanges.

I won't close the issue yet as I would like to add the diffs of the PSycloned adjoint tests so we have a better understanding as to what is going on.

**Comment by mo-joshuacolclough** [comment]
> Hmm, is the metadata on the adjoint kernel correct?

Apologies I somehow missed this - I checked the adjoint kernel metadata against the linear metadata and both seem correct to me.

**Comment by DrTVockerodtMO** [comment]
Here is the output before and after the config change in the file I originally showed:

```
--- atlt_vorticity_advection_alg_mod_psy_before.f90	2025-02-27 10:51:11.549615000 +0000
+++ atlt_vorticity_advection_alg_mod_psy_after.f90	2025-02-27 10:28:33.816454000 +0000
@@ -215,21 +215,21 @@
       ! Set-up all of the loop bounds
       !
       loop0_start = 1
-      loop0_stop = r_u_proxy%vspace%get_last_dof_annexed()
+      loop0_stop = r_u_proxy%vspace%get_last_dof_owned()
       loop1_start = 1
-      loop1_stop = r_u_input_proxy%vspace%get_last_dof_annexed()
+      loop1_stop = r_u_input_proxy%vspace%get_last_dof_owned()
       loop2_start = 1
-      loop2_stop = wind_proxy%vspace%get_last_dof_annexed()
+      loop2_stop = wind_proxy%vspace%get_last_dof_owned()
       loop3_start = 1
-      loop3_stop = wind_input_proxy%vspace%get_last_dof_annexed()
+      loop3_stop = wind_input_proxy%vspace%get_last_dof_owned()
       loop4_start = 1
-      loop4_stop = vorticity_proxy%vspace%get_last_dof_annexed()
+      loop4_stop = vorticity_proxy%vspace%get_last_dof_owned()
       loop5_start = 1
-      loop5_stop = vorticity_input_proxy%vspace%get_last_dof_annexed()
+      loop5_stop = vorticity_input_proxy%vspace%get_last_dof_owned()
       loop6_start = 1
-      loop6_stop = ls_wind_proxy%vspace%get_last_dof_annexed()
+      loop6_stop = ls_wind_proxy%vspace%get_last_dof_owned()
       loop7_start = 1
-      loop7_stop = ls_vorticity_proxy%vspace%get_last_dof_annexed()
+      loop7_stop = ls_vorticity_proxy%vspace%get_last_dof_owned()
       loop8_start = 1
       loop8_stop = ncolour
       loop9_start = 1
@@ -362,6 +362,7 @@
       !
       ! End of set dirty/clean section for above loop(s)
       !
+      CALL r_u_proxy%halo_exchange(depth=1)
       CALL wind_proxy%halo_exchange(depth=1)
       CALL vorticity_proxy%halo_exchange(depth=1)
       CALL ls_wind_proxy%halo_exchange(depth=1)
@@ -668,6 +669,12 @@
       IF (r_u_proxy%is_dirty(depth=1)) THEN
         CALL r_u_proxy%halo_exchange(depth=1)
       END IF
+      IF (wind_proxy%is_dirty(depth=1)) THEN
+        CALL wind_proxy%halo_exchange(depth=1)
+      END IF
+      IF (vorticity_proxy%is_dirty(depth=1)) THEN
+        CALL vorticity_proxy%halo_exchange(depth=1)
+      END IF
       IF (ls_wind_proxy%is_dirty(depth=1)) THEN
         CALL ls_wind_proxy%halo_exchange(depth=1)
       END IF
```

It looks like we are capturing the needed halo exchanges now.

**Comment by DrTVockerodtMO** [comment]
This was causing some floating invalid errors when running the adjoint code, and it helped uncover a bug in our patches. Essentially the metadata for kernels with multiple code implementations inherits from the forward code, which makes sense because sometimes multiple routines in the code can correspond to helper functions etc. where one would not be able to derive the metadata. I made an issue previously about it here: https://github.com/stfc/PSyclone/issues/2495.

What I focused more on was the naming of the adjoint kernel, but in fact it's the entire metadata that is inherited. This includes, for instance, having `GH_READ` for fields that should be `GH_WRITE` or `GH_INC`! The main difference there is that the vital halo exchanges are not called when running the application which cause the floating point errors. After modifying the kernel metadata to be correct I now get these particular kernels with floating point errors to pass! I'll include the patch changes on my LFRic apps branch.

**Comment by mo-joshuacolclough** [comment]
Thanks for all the help on this - I've been doing some testing with this with `COMPUTE_ANNEXED_DOFS=false`; this works fine for kernels where one of the fields is `GH_INC`, as halo exchanges are added for the correct fields.

However for kernels where there is only `GH_WRITE`, there are no halo exchanges. Because `setval_random` populates the field until the end of the `get_last_owned_dof()`, when accessing the cell dofmaps in the following kernel calls for W2/W1/W0 function spaces, the kernel accesses annexed dofs which are not set by `setval_random` (nor should they be - #2909) so you end up with floating invalid errors.

The problem is that even if annexed dofs were set, but dirty (e.g randomized), on the local process this would be problematic for a matrix operation between W3 and W2 for example, because this "dirty" annexed dof would result in an incorrect answer - for the correct answer these dofs have to be clean. To me either PSyclone needs some way to check `annexed` dofs are dirty (rather than halo dofs), or `setval_random` is a special case which needs to perform a halo exchange to uphold the assumption that annexed dofs are always clean. Wondering thoughts on this...

---

Kernel metadata for the problematic kernel:

```fortran
  type, public, extends(kernel_type) :: dg_matrix_vector_kernel_type
    private
    type(arg_type) :: meta_args(3) = (/                                       &
         arg_type(GH_FIELD,    GH_REAL, GH_WRITE, ANY_DISCONTINUOUS_SPACE_1), &
         arg_type(GH_FIELD,    GH_REAL, GH_READ,  ANY_SPACE_1),               &
         arg_type(GH_OPERATOR, GH_REAL, GH_READ,  ANY_DISCONTINUOUS_SPACE_1,  &
                                                  ANY_SPACE_1)                &
         /)
    integer :: operates_on = CELL_COLUMN
  end type
```

Generated PSyclone invoke:

- `invoke_0` - Initial setup of the fields with setval_random and setval_X.
- `invoke_tl` - First kernel call.

```fortran
    SUBROUTINE invoke_0(lhs, lhs_input, x, x_input, matrix)
      USE setop_random_kernel_mod, ONLY: setop_random_kernel_code
      USE mesh_mod, ONLY: mesh_type
      TYPE(field_type), intent(in) :: lhs, lhs_input, x, x_input
      TYPE(operator_type), intent(in) :: matrix
      INTEGER(KIND=i_def) cell
      INTEGER(KIND=i_def) df
      INTEGER(KIND=i_def) loop4_start, loop4_stop
      INTEGER(KIND=i_def) loop3_start, loop3_stop
      INTEGER(KIND=i_def) loop2_start, loop2_stop
      INTEGER(KIND=i_def) loop1_start, loop1_stop
      INTEGER(KIND=i_def) loop0_start, loop0_stop
      INTEGER(KIND=i_def) nlayers_matrix
      REAL(KIND=r_def), pointer, dimension(:,:,:) :: matrix_local_stencil => null()
      TYPE(operator_proxy_type) matrix_proxy
      REAL(KIND=r_def), pointer, dimension(:) :: x_input_data => null()
      REAL(KIND=r_def), pointer, dimension(:) :: x_data => null()
      REAL(KIND=r_def), pointer, dimension(:) :: lhs_input_data => null()
      REAL(KIND=r_def), pointer, dimension(:) :: lhs_data => null()
      TYPE(field_proxy_type) lhs_proxy, lhs_input_proxy, x_proxy, x_input_proxy
      INTEGER(KIND=i_def) ndf_aspc1_lhs, undf_aspc1_lhs, ndf_aspc1_lhs_input, undf_aspc1_lhs_input, ndf_aspc1_x, undf_aspc1_x, &
&ndf_aspc1_x_input, undf_aspc1_x_input, ndf_aspc1_matrix, ndf_aspc2_matrix
      INTEGER(KIND=i_def) max_halo_depth_mesh
      TYPE(mesh_type), pointer :: mesh => null()
      !
      ! Initialise field and/or operator proxies
      !
      lhs_proxy = lhs%get_proxy()
      lhs_data => lhs_proxy%data
      lhs_input_proxy = lhs_input%get_proxy()
      lhs_input_data => lhs_input_proxy%data
      x_proxy = x%get_proxy()
      x_data => x_proxy%data
      x_input_proxy = x_input%get_proxy()
      x_input_data => x_input_proxy%data
      matrix_proxy = matrix%get_proxy()
      matrix_local_stencil => matrix_proxy%local_stencil
      !
      ! Initialise number of layers
      !
      nlayers_matrix = matrix_proxy%fs_from%get_nlayers()
      !
      ! Create a mesh object
      !
      mesh => lhs_proxy%vspace%get_mesh()
      max_halo_depth_mesh = mesh%get_halo_depth()
      !
      ! Initialise number of DoFs for aspc1_lhs
      !
      ndf_aspc1_lhs = lhs_proxy%vspace%get_ndf()
      undf_aspc1_lhs = lhs_proxy%vspace%get_undf()
      !
      ! Initialise number of DoFs for aspc1_lhs_input
      !
      ndf_aspc1_lhs_input = lhs_input_proxy%vspace%get_ndf()
      undf_aspc1_lhs_input = lhs_input_proxy%vspace%get_undf()
      !
      ! Initialise number of DoFs for aspc1_x
      !
      ndf_aspc1_x = x_proxy%vspace%get_ndf()
      undf_aspc1_x = x_proxy%vspace%get_undf()
      !
      ! Initialise number of DoFs for aspc1_x_input
      !
      ndf_aspc1_x_input = x_input_proxy%vspace%get_ndf()
      undf_aspc1_x_input = x_input_proxy%vspace%get_undf()
      !
      ! Initialise number of DoFs for aspc1_matrix
      !
      ndf_aspc1_matrix = matrix_proxy%fs_to%get_ndf()
      !
      ! Initialise number of DoFs for aspc2_matrix
      !
      ndf_aspc2_matrix = matrix_proxy%fs_from%get_ndf()
      !
      ! Set-up all of the loop bounds
      !
      loop0_start = 1
      loop0_stop = lhs_proxy%vspace%get_last_dof_owned()
      loop1_start = 1
      loop1_stop = lhs_input_proxy%vspace%get_last_dof_owned()
      loop2_start = 1
      loop2_stop = x_proxy%vspace%get_last_dof_owned()
      loop3_start = 1
      loop3_stop = x_input_proxy%vspace%get_last_dof_owned()
      loop4_start = 1
      loop4_stop = mesh%get_last_halo_cell(1)
      !
      ! Call kernels and communication routines
      !
      DO df = loop0_start, loop0_stop, 1
        ! Built-in: setval_random (fill a real-valued field with pseudo-random numbers)
        CALL RANDOM_NUMBER(lhs_data(df))
      END DO
      !
      ! Set halos dirty/clean for fields modified in the above loop
      !
      CALL lhs_proxy%set_dirty()
      !
      DO df = loop1_start, loop1_stop, 1
        ! Built-in: setval_X (set a real-valued field equal to another such field)
        lhs_input_data(df) = lhs_data(df)
      END DO
      !
      ! Set halos dirty/clean for fields modified in the above loop
      !
      CALL lhs_input_proxy%set_dirty()
      !
      DO df = loop2_start, loop2_stop, 1
        ! Built-in: setval_random (fill a real-valued field with pseudo-random numbers)
        CALL RANDOM_NUMBER(x_data(df))
      END DO
      !
      ! Set halos dirty/clean for fields modified in the above loop
      !
      CALL x_proxy%set_dirty()
      !
      DO df = loop3_start, loop3_stop, 1
        ! Built-in: setval_X (set a real-valued field equal to another such field)
        x_input_data(df) = x_data(df)
      END DO
      !
      ! Set halos dirty/clean for fields modified in the above loop
      !
      CALL x_input_proxy%set_dirty()
      !
      DO cell = loop4_start, loop4_stop, 1
        CALL setop_random_kernel_code(cell, nlayers_matrix, matrix_proxy%ncell_3d, matrix_local_stencil, ndf_aspc1_matrix, &
&ndf_aspc2_matrix)
      END DO
      !
    END SUBROUTINE invoke_0
    SUBROUTINE invoke_tl(lhs, x, matrix)
      USE dg_matrix_vector_kernel_mod, ONLY: dg_matrix_vector_code
      USE mesh_mod, ONLY: mesh_type
      TYPE(field_type), intent(in) :: lhs, x
      TYPE(operator_type), intent(in) :: matrix
      INTEGER(KIND=i_def) cell
      INTEGER(KIND=i_def) loop0_start, loop0_stop
      INTEGER(KIND=i_def) nlayers_lhs
      REAL(KIND=r_def), pointer, dimension(:,:,:) :: matrix_local_stencil => null()
      TYPE(operator_proxy_type) matrix_proxy
      REAL(KIND=r_def), pointer, dimension(:) :: x_data => null()
      REAL(KIND=r_def), pointer, dimension(:) :: lhs_data => null()
      TYPE(field_proxy_type) lhs_proxy, x_proxy
      INTEGER(KIND=i_def), pointer :: map_adspc1_lhs(:,:) => null(), map_aspc1_x(:,:) => null()
      INTEGER(KIND=i_def) ndf_adspc1_lhs, undf_adspc1_lhs, ndf_aspc1_x, undf_aspc1_x
      INTEGER(KIND=i_def) max_halo_depth_mesh
      TYPE(mesh_type), pointer :: mesh => null()
      !
      ! Initialise field and/or operator proxies
      !
      lhs_proxy = lhs%get_proxy()
      lhs_data => lhs_proxy%data
      x_proxy = x%get_proxy()
      x_data => x_proxy%data
      matrix_proxy = matrix%get_proxy()
      matrix_local_stencil => matrix_proxy%local_stencil
      !
      ! Initialise number of layers
      !
      nlayers_lhs = lhs_proxy%vspace%get_nlayers()
      !
      ! Create a mesh object
      !
      mesh => lhs_proxy%vspace%get_mesh()
      max_halo_depth_mesh = mesh%get_halo_depth()
      !
      ! Look-up dofmaps for each function space
      !
      map_adspc1_lhs => lhs_proxy%vspace%get_whole_dofmap()
      map_aspc1_x => x_proxy%vspace%get_whole_dofmap()
      !
      ! Initialise number of DoFs for adspc1_lhs
      !
      ndf_adspc1_lhs = lhs_proxy%vspace%get_ndf()
      undf_adspc1_lhs = lhs_proxy%vspace%get_undf()
      !
      ! Initialise number of DoFs for aspc1_x
      !
      ndf_aspc1_x = x_proxy%vspace%get_ndf()
      undf_aspc1_x = x_proxy%vspace%get_undf()
      !
      ! Set-up all of the loop bounds
      !
      loop0_start = 1
      loop0_stop = mesh%get_last_edge_cell()
      !
      ! Call kernels and communication routines
      !
      DO cell = loop0_start, loop0_stop, 1
        CALL dg_matrix_vector_code(cell, nlayers_lhs, lhs_data, x_data, matrix_proxy%ncell_3d, matrix_local_stencil, &
&ndf_adspc1_lhs, undf_adspc1_lhs, map_adspc1_lhs(:,cell), ndf_aspc1_x, undf_aspc1_x, map_aspc1_x(:,cell))
      END DO
      !
      ! Set halos dirty/clean for fields modified in the above loop
      !
      CALL lhs_proxy%set_dirty()
      !
      !
    END SUBROUTINE invoke_tl
```

**Comment by mo-joshuacolclough** [comment]
Using `COMPUTE_ANNEXED_DOFS=false` breaks the assumption that annexed dofs are always clean (because built-ins loop over `owned` dofs), so we end up with adjoint failures later down the line when built-ins are used, in particular when algorithms initialise local fields with built-ins, prior to invoking kernels that use annexed dofs.

Manually changing the psy-layer code in those cases to loop into `get_last_dof_annexed` fixes the adjoint. It's likely we have to switch `COMPUTE_ANNEXED_DOFS=true` back on, but this shows turning this off breaks the assumptions around annexed dofs. I've opened #2992

**Comment by DrTVockerodtMO** [comment]
> Using `COMPUTE_ANNEXED_DOFS=false` breaks the assumption that annexed dofs are always clean (because built-ins loop over `owned` dofs), so we end up with adjoint failures later down the line when built-ins are used, in particular when algorithms initialise local fields with built-ins, prior to invoking kernels that use annexed dofs with `GH_WRITE` as opposed to `GH_INC`.
> 
> Manually changing the psy-layer code for the built-ins in those cases to loop into `get_last_dof_annexed` fixes the adjoint. It's likely we have to switch `COMPUTE_ANNEXED_DOFS=true` back on, but this shows turning this off breaks the assumptions around annexed dofs. I've opened [#2992](https://github.com/stfc/PSyclone/issues/2992)

Turning `COMPUTE_ANNEXED_DOFS=true` back on will adjust the loop indices but I think it would still be missing the halo exchanges needed to properly setup some of the adjoint tests (as well as not setting the halo to dirty for when we do need a halo exchange). We have to be careful here.

**Comment by DrTVockerodtMO** [comment]
Might it be fair to say that if `COMPUTE_ANNEXED_DOFS=false` is assuming that the annexed dofs are clean, that that should be what gets modified?

**Comment by DrTVockerodtMO** [comment]
> The metadata for matrix-vector is:
> 
> ```
>   type, public, extends(kernel_type) :: matrix_vector_kernel_type
>     private
>     type(arg_type) :: meta_args(3) = (/                                    &
>          arg_type(GH_FIELD,    GH_REAL, GH_INC,  ANY_SPACE_1),             &
>          arg_type(GH_FIELD,    GH_REAL, GH_READ, ANY_SPACE_2),             &
>          arg_type(GH_OPERATOR, GH_REAL, GH_READ, ANY_SPACE_1, ANY_SPACE_2) &
>          /)
>     integer :: operates_on = CELL_COLUMN
>   end type
> ```
> 
> so it _does_ perform GH_INC on a field on a continuous (or potentially continuous) function space. This will require annexed dofs to be clean on entry. This would normally trigger a halo exchange but, when using `COMPUTE_ANNEXED_DOFS` (s the MO do by default), the annexed dofs are always assumed to be clean and no halo exchange is added. The annexed dofs are made clean by ensuring that the `setval_*` kernels always (redundantly) update them - note that this will still leave parts of the L1 halo dirty. However, because matrix-vector operates on cells, it will still be accessing 'halo' dofs, even if the results are never used. Therefore, if those 'halo' dofs contain invalid data, we will get f.p. exceptions. The recommended way to ensure this doesn't happen is to always perform redundant computation to the L1 halo for _all_ `setval_` kernels. Once we see the generated PSy-layer, we'll be able to see definitively whether this is being done.

This solution may be relevant for what @mo-joshuacolclough is describing.

**Comment by mo-joshuacolclough** [comment]
Thanks @DrTVockerodtMO, it's possible fixing #2992 would solve our problems.

**Comment by arporter** [comment]
Well done. #2909 won't fix anything for you - it will just ensure that `COMPUTE_ANNEXED_DOFS` isn't set to True when you call `setval_random`. However, #2994 fixes the bug you found with missing halo exchanges. This will solve at least some of your problems. The branch is `2992_halo_exch_op_kernel` if you'd like to test it.

**Comment by mo-joshuacolclough** [comment]
Ah I see, thank you for the clarification. Thank you again for all the work on #2994. I will test it on our system. Cheers

**Comment by DrTVockerodtMO** [comment]
> Update: For now we have cobbled together an optimisation script that forces a halo exchange after calls to `setval_random`, which has solved the adjoint test issue until [#2909](https://github.com/stfc/PSyclone/issues/2909) can be used.

I believe it solves some of the issues but without the halo exchanges for `GH_WRITE` in the forward/adjoint kernels there are still adjoint correctness failures in some of the tests.
